cpp/include/cudf/scalar/scalar.hpp

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/types.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/utilities/traits.hpp>
+
+#include <rmm/device_buffer.hpp>
+#include <rmm/device_scalar.hpp>
+#include <rmm/thrust_rmm_allocator.h>
+#include <rmm/mr/device_memory_resource.hpp>
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace cudf {
+
+/**
+ * @brief An owning class to represent a singular value
+ * 
+ * A scalar is a singular value of any of the supported datatypes in cudf. 
+ * Classes derived from this class are used to represent a scalar. Objects of
+ * derived classes should be upcasted to this class while passing to an
+ * external libcudf API.
+ */
+class scalar {
+ public:
+  scalar() = default;
```

Comment:
This should be protected or deleted, yes? We don't want people to be able to directly instantiate the `scalar` base class.